### ISSUE TITLE
⚠️ Version 2.0

### ISSUE THREAD

**Issue by bwp91** [bwp91/homebridge-platform-wemo#206]
### Why 2.0.0?
* To mark a change in plugin maintainer which is now me (bwp91).
* To make it easy to revert back to v1 branch in case of any problems.

---

### Configuration Changes
* NEW: `debug` setting (default:false) when true will add non-warning/error messages to the HB log.
* NEW: Control over the Wemo Maker type is now through the plugin configuration.
* CHANGE: `discovery` is now `disableDiscovery` (default:false) when true will disable discovery.

Please see [the wiki](https://github.com/bwp91/homebridge-platform-wemo/wiki/Configuration-(v2.0)) for templates.

---

### Other Changes
* Reorganisation of the plugin settings screen.
* Removal of `configurationRequestHandler`. The only reason I saw for this was to change the Wemo Maker type (Switch <-> GarageDoor) which can now be done through the plugin config.
* Code refactoring (personal preferences) to enable me to maintain this plugin more easily.
* **Apart from these changes above, my plan for v2.0.0 is to function exactly the same as v1.6.1.** I then plan to implement new features in subsequent versions.

---

### 🟢 What's Working
From user feedback:
* Wemo Bulb (via Link)
* Wemo Motion
* Wemo Light Switch
* Wemo Switch
* Wemo Insight
* Wemo Maker

---

### 🟠 What Needs Testing
* Wemo Dimmer
* Wemo NetCam (only as motion sensor)

---

If you're happy to try it out please do give feedback by commenting on this issue - I'm interested to know what **is working** and what **isn't working**. I shall keep the above updated from user feedback.

Thanks
Ben

**Comment by donavanbecker** [comment]
I did get an error on the latest beta.

**Comment by donavanbecker** [comment]
@bwp91 2.0.0-4 is working now.

**Comment by vsidlu** [comment]
Hi!
I got two errors - 

[10/20/2020, 19:44:03] [WeMo Platform] [WeMo Link] an error occurred requesting subdevices - TypeError: client.getsubDevices is not a function
--
at wemoPlatform.initialiseDevice (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-platform.js:111:18)
at module.exports.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-platform.js:68:14)
at /usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-client/index.js:50:14
at Parser.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:304:18)
at Parser.emit (events.js:310:20)
at SAXParser.onclosetag (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:262:26)
at emit (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/sax/lib/sax.js:624:35)
at emitNode (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/sax/lib/sax.js:629:5)
at closeTag (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/sax/lib/sax.js:889:7)
at SAXParser.write (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/sax/lib/sax.js:1436:13)
at Parser.exports.Parser.Parser.parseString (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:323:31)
at Parser.parseString (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:5:59)
at Object.exports.parseString (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:369:19)
at IncomingMessage.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-client/client.js:67:16)
at IncomingMessage.emit (events.js:322:22)
at endReadableNT (_stream_readable.js:1187:12)

and

[10/20/2020, 19:44:04] TypeError: Cannot read property 'noMotionTimer' of undefined
--
at deviceMotion.updateMotionDetected (/usr/lib/node_modules/homebridge-platform-wemo/lib/device/motion.js:24:32)
at module.exports.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/lib/device/motion.js:14:54)
at module.exports.emit (events.js:310:20)
at Object.BinaryState (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-client/client.js:377:12)
at /usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-client/client.js:420:22
at Parser.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:304:18)
at Parser.emit (events.js:310:20)
at SAXParser.onclosetag (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:262:26)
at emit (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/sax/lib/sax.js:624:35)
at emitNode (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/sax/lib/sax.js:629:5)
at closeTag (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/sax/lib/sax.js:889:7)
at SAXParser.write (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/sax/lib/sax.js:1436:13)
at Parser.exports.Parser.Parser.parseString (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:323:31)
at Parser.parseString (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:5:59)
at Object.exports.parseString (/usr/lib/node_modules/homebridge-platform-wemo/node_modules/xml2js/lib/parser.js:369:19)
at module.exports.WemoClient.handleCallback (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-client/client.js:416:10)

**Comment by bwp91** [comment]
@vsidlu these should be fixed in 2.0.0-5
and @donavanbecker 
thank you both for your reports!

**Comment by vsidlu** [comment]
you are wellcome - another one - 


[WeMo Link] could not be initialised as TypeError: Cannot read property 'displayName' of undefined
--
at wemoPlatform.initialiseDevice (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-platform.js:205:65)
at module.exports.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-platform.js:68:14)
at /usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-client/index.js:50:14
at Parser.<anonymous> (/usr/lib/node_modules/xml2js/lib/parser.js:304:18)
at Parser.emit (events.js:310:20)
at SAXParser.onclosetag (/usr/lib/node_modules/xml2js/lib/parser.js:262:26)
at emit (/usr/lib/node_modules/sax/lib/sax.js:624:35)
at emitNode (/usr/lib/node_modules/sax/lib/sax.js:629:5)
at closeTag (/usr/lib/node_modules/sax/lib/sax.js:889:7)
at SAXParser.write (/usr/lib/node_modules/sax/lib/sax.js:1436:13)
at Parser.exports.Parser.Parser.parseString (/usr/lib/node_modules/xml2js/lib/parser.js:323:31)
at Parser.parseString (/usr/lib/node_modules/xml2js/lib/parser.js:5:59)
at Object.exports.parseString (/usr/lib/node_modules/xml2js/lib/parser.js:369:19)
at IncomingMessage.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-client/client.js:67:16)
at IncomingMessage.emit (events.js:322:22)
at endReadableNT (_stream_readable.js:1187:12).

**Comment by vsidlu** [comment]
you are wellcome :) - another one - wemo link still not working

[WeMo Link] could not be initialised as TypeError: Cannot read property 'displayName' of undefined
--
at wemoPlatform.initialiseDevice (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-platform.js:205:65)
at module.exports.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-platform.js:68:14)
at /usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-client/index.js:50:14
at Parser.<anonymous> (/usr/lib/node_modules/xml2js/lib/parser.js:304:18)
at Parser.emit (events.js:310:20)
at SAXParser.onclosetag (/usr/lib/node_modules/xml2js/lib/parser.js:262:26)
at emit (/usr/lib/node_modules/sax/lib/sax.js:624:35)
at emitNode (/usr/lib/node_modules/sax/lib/sax.js:629:5)
at closeTag (/usr/lib/node_modules/sax/lib/sax.js:889:7)
at SAXParser.write (/usr/lib/node_modules/sax/lib/sax.js:1436:13)
at Parser.exports.Parser.Parser.parseString (/usr/lib/node_modules/xml2js/lib/parser.js:323:31)
at Parser.parseString (/usr/lib/node_modules/xml2js/lib/parser.js:5:59)
at Object.exports.parseString (/usr/lib/node_modules/xml2js/lib/parser.js:369:19)
at IncomingMessage.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/lib/wemo-client/client.js:67:16)
at IncomingMessage.emit (events.js:322:22)
at endReadableNT (_stream_readable.js:1187:12).

**Comment by bwp91** [comment]
@vsidlu should be fixed now v2.0.0-6

**Comment by vsidlu** [comment]
[10/20/2020, 21:06:01] TypeError: Cannot read property 'setDeviceStatus' of undefined
--
at deviceLink.setSwitchState (/usr/lib/node_modules/homebridge-platform-wemo/lib/device/link.js:151:17)
at Characteristic.<anonymous> (/usr/lib/node_modules/homebridge-platform-wemo/lib/device/link.js:24:44)
at Characteristic.emit (events.js:310:20)
at Characteristic.EventEmitter.emit (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/EventEmitter.ts:42:22)
at Characteristic._this.setValue (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:652:12)
at /usr/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:1437:24
at Array.forEach (<anonymous>)
at Bridge.Accessory._this._handleSetCharacteristics (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Accessory.ts:1279:10)
at HAPServer.emit (events.js:310:20)
at HAPServer.EventEmitter.emit (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/EventEmitter.ts:42:22)
at HAPServer._this._handleCharacteristics (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/HAPServer.ts:812:12)
at IncomingMessage.<anonymous> (/usr/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/HAPServer.ts:280:24)
at IncomingMessage.emit (events.js:310:20)
at endReadableNT (_stream_readable.js:1187:12)
at processTicksAndRejections (internal/process/task_queues.js:84:21)

**Comment by bwp91** [comment]
@vsidlu 2.0.0-7!

**Comment by vsidlu** [comment]
Thanks! Now it is working! I got also warm and cold color option for bulbs. Did not see it in previous versions.
Now plese help with json for wemo maker type garage door.

**Comment by bwp91** [comment]
> Thanks! Now it is working! I got also warm and cold color option for bulbs. Did not see it in previous versions.

Can’t take credit for that, I haven’t *added* any new features to this plugin!

> Now plese help with json for wemo maker type garage door.

@vsidlu are you on the Homebridge discord channel?

**Comment by vsidlu** [comment]
@bwp91 yes I am!

**Comment by julien94800** [comment]
Hello,
Just updated to new version. I am using this plugin with two maker set as garage door and it seems to work as expected after manually modifying the config as described. 
Thanks for your work!!
What would be useful if possible is to be able to set makers as garage door with the gui! Currently, it seems to be only feasible manually

Once again, thanks!!

**Comment by Nastras** [comment]
I have also just done the update. In my case the garage door is only shown as a switch with contact sensor in HomeKit. Before with the old version it was shown as a garage door. I have adapted the Config. 

Any idea why?

"platforms": [
{
"platform": "BelkinWeMo",
"name": "WeMo Platform",
"doorOpenTimer": 20,
"manualDevices": [],
"disableDiscovery": false,
"discoveryInterval": 30,
"makerTypes": [{"serialNumber": "XXXXXX", "makerType": "garageDoor"}],
"wemoClient": {}
}
]
}

Thanks!

**Comment by Nastras** [comment]
It looks like the plugin does not consider this option:
{"serialNumber": "XXXXXXX", "makerType": "**garageDoor**"}

@julien94800 is it displayed as a garage door?

**Comment by permaMonk** [comment]
Last update I get this:
[10/25/2020, 15:42:40] [Belkin WeMo] Plugin has finished initialising. Finding devices to add to Homebridge.
[10/25/2020, 15:42:40] [Belkin WeMo] An error occured loading a device - Error: connect ENETUNREACH 192.168.178.36:49153 - Local (0.0.0.0:0)
    at internalConnect (net.js:924:16)
    at defaultTriggerAsyncIdScope (internal/async_hooks.js:364:12)
    at net.js:1012:9
    at processTicksAndRejections (internal/process/task_queues.js:79:11) {
  errno: 'ENETUNREACH',
  code: 'ENETUNREACH',
  syscall: 'connect',
  address: '192.168.178.36',
  port: 49153
}
Unhandled rejection Error: No sockets available, cannot start.
    at SsdpClient.SSDP._createSockets (/usr/local/lib/node_modules/homebridge-platform-wemo/node_modules/node-ssdp/lib/index.js:186:11)
    at SsdpClient.SSDP._start (/usr/local/lib/node_modules/homebridge-platform-wemo/node_modules/node-ssdp/lib/index.js:229:10)
    at /usr/local/lib/node_modules/homebridge-platform-wemo/node_modules/node-ssdp/lib/client.js:60:10
    at Promise._execute (/usr/local/lib/node_modules/homebridge-platform-wemo/node_modules/bluebird/js/release/debuggability.js:384:9)
    at Promise._resolveFromExecutor (/usr/local/lib/node_modules/homebridge-platform-wemo/node_modules/bluebird/js/release/promise.js:518:18)
    at new Promise (/usr/local/lib/node_modules/homebridge-platform-wemo/node_modules/bluebird/js/release/promise.js:103:10)
    at SsdpClient.start (/usr/local/lib/node_modules/homebridge-platform-wemo/node_modules/node-ssdp/lib/client.js:54:10)
    at SsdpClient.search (/usr/local/lib/node_modules/homebridge-platform-wemo/node_modules/node-ssdp/lib/client.js:82:17)
    at wemoPlatform.discoverDevices (/usr/local/lib/node_modules/homebridge-platform-wemo/lib/wemo-platform.js:124:22)
    at wemoPlatform.wemoSetup (/usr/local/lib/node_modules/homebridge-platform-wemo/lib/wemo-platform.js:101:14)
    at HomebridgeAPI.<anonymous> (/usr/local/lib/node_modules/homebridge-platform-wemo/lib/wemo-platform.js:54:44)
    at HomebridgeAPI.emit (events.js:314:20)
    at HomebridgeAPI.signalFinished (/usr/local/lib/node_modules/homebridge/src/api.ts:251:10)
    at Server.start (/usr/local/lib/node_modules/homebridge/src/server.ts:158:14)
    at cli (/usr/local/lib/node_modules/homebridge/src/cli.ts:80:10)
    at Object.<anonymous> (/usr/local/lib/node_modules/homebridge/bin/homebridge:17:22)

**Comment by Nastras** [comment]
Error found. I have here: "serialNumber": "XXXXXX", the Mac address that is displayed in the Journalctl because I assume that this is the serial number.

I have now entered the serial number on the device and now he recognizes it as a garage door.

@bwp91 It would be great if you can display the serial number in addition to the Mac address of the found devices when you start the Homebridge.

Thanks for your work!

**Comment by julien94800** [comment]
Hello @Nastras 
Sorry did not had time to answer. Yes everything is working great and shown as garage door. Seems to be the case for you too now so perfect

**Comment by bwp91** [comment]
Hi @julien94800 

> What would be useful if possible is to be able to set makers as garage door with the gui! Currently, it seems to be only feasible manually

This is definitely an option, I was just trying to not "clutter" the plugin settings screen with options that are potentially used just once. But I can look into this. Glad the garage is working for you.

Hi @Nastras 

> It would be great if you can display the serial number in addition to the Mac address of the found devices when you start the Homebridge.

This is definitely something I can do. Glad the garage is working for you now!

Hi @permaMonk 

Haven't seen this before. Does it occur again if you restart Homebridge? I'll look into this.


Thanks to all for your feedback

**Comment by permaMonk** [comment]
Hi bwp91.

>>Haven't seen this before. Does it occur again if you restart Homebridge? I'll look into this.

Yes I´ve done several restarts yesterady. But this day I´ve tried another restart and the HB and logfile looks really fine. And the Wemo Plug works also again. I´m sorry for trouble I made. Maybe it is due to an interference with another plugin (logic) I recently installed. Because there were some problems with it, I removed its config entry. Maybe? Thanks again for all your commitment.
